circom/aes256.go

```diff
@@ -0,0 +1,318 @@
+	var newState [16]frontend.Variable
+	for i := 0; i < 4; i++ {
+		newState[i] = aes.variableXor(state[i], expandedKey[from+(4*i)], 8)
+		newState[4+i] = aes.variableXor(state[4+i], expandedKey[from+(4*i)+1], 8)
+		newState[8+i] = aes.variableXor(state[8+i], expandedKey[from+(4*i)+2], 8)
+		newState[12+i] = aes.variableXor(state[12+i], expandedKey[from+(4*i)+3], 8)
+	}
+	return newState
+}
+
+// different re-arrangement of variables
+func (aes *AES256) addRoundKey2(state [16]frontend.Variable, expandedKey [240]frontend.Variable, from int) [16]frontend.Variable {
+	var newState [16]frontend.Variable
+	ctr := 0
+	for i := 0; i < 4; i++ {
+		newState[i] = aes.variableXor(state[i], expandedKey[from+ctr], 8)
+		newState[4+i] = aes.variableXor(state[4+i], expandedKey[from+ctr+1], 8)
+		newState[8+i] = aes.variableXor(state[8+i], expandedKey[from+ctr+2], 8)
+		newState[12+i] = aes.variableXor(state[12+i], expandedKey[from+ctr+3], 8)
+		ctr += 4
+	}
+	return newState
+}
+
+// xor on bits of two frontend.Variables
+func (aes *AES256) variableXor(a frontend.Variable, b frontend.Variable, size int) frontend.Variable {
+	bitsA := aes.api.ToBinary(a, size)
+	bitsB := aes.api.ToBinary(b, size)
+	x := make([]frontend.Variable, size)
+	for i := 0; i < size; i++ {
+		x[i] = aes.api.Xor(bitsA[i], bitsB[i])
+	}
+	return aes.api.FromBinary(x...)
+}
+
+// expands 32 byte key to 240 byte output
+func (aes *AES256) expandKey(key [32]frontend.Variable, sbox0 [256]frontend.Variable, RCon [30]frontend.Variable) [240]frontend.Variable {
+
+	var expand [240]frontend.Variable
+	i := 0
+
+	for i < 32 {
+		expand[i] = key[i]
+		expand[i+1] = key[i+1]
+		expand[i+2] = key[i+2]
+		expand[i+3] = key[i+3]
+
+		i += 4
+	}
+
+	for i < 240 {
+		t0 := expand[i-4]
+		t1 := expand[i-3]
+		t2 := expand[i-2]
+		t3 := expand[i-1]
+
+		if i%32 == 0 {
+			// t = subw(rotw(t)) ^ (uint32(powx[i/nb-1]) << 24)
+
+			// rotation
+			t0, t1, t2, t3 = t1, t2, t3, t0
+
+			// subwords
+			t0 = aes.subw(sbox0, t0)
+			t1 = aes.subw(sbox0, t1)
+			t2 = aes.subw(sbox0, t2)
+			t3 = aes.subw(sbox0, t3)
+			t0 = aes.variableXor(t0, RCon[i/32], 8)
+		} else if i%32 == 16 {
+			t0 = aes.subw(sbox0, t0)
+			t1 = aes.subw(sbox0, t1)
+			t2 = aes.subw(sbox0, t2)
+			t3 = aes.subw(sbox0, t3)
+		}
+
+		expand[i] = aes.variableXor(expand[i-32], t0, 8)
+		expand[i+1] = aes.variableXor(expand[i-32+1], t1, 8)
+		expand[i+2] = aes.variableXor(expand[i-32+2], t2, 8)
+		expand[i+3] = aes.variableXor(expand[i-32+3], t3, 8)
+
+		i += 4
+	}
+
+	return expand
+}
+
+// substitute word with naive lookup of sbox
+func (aes *AES256) subw(sbox [256]frontend.Variable, a frontend.Variable) frontend.Variable {
+	out := frontend.Variable(0)
+	for j := 0; j < 256; j++ {
+		out = aes.api.Add(out, aes.api.Mul(aes.api.IsZero(aes.api.Sub(a, j)), sbox[j]))
+		// api.Cmp instead of api.Sub works but is inefficient
+	}
```

Comment:
I think the execution efficiency of the entire circuit is not something that needs to be considered. Because this part of the calculation actually occurs off-chain and is only executed once per epoch. As long as the overall calculation time doesn't take several days, it should be fine.

Reply:
Got it, thanks.